hera_cal/utils.py

```diff
@@ -731,3 +731,114 @@ def synthesize_ant_flags(flags, threshold=0.0):
             ant_flags[ap] = (ant_Nflag[ap] / ant_Nvis[ap]) > threshold
 
     return ant_flags
+
+
+def chisq(data, model, data_wgts, gains=None, gain_flags=None, by_antpol=False,
+          chisq=None, nObs=None, chisq_per_ant=None, nObs_per_ant=None):
+    """Computes chi^2 defined as:
+
+    chi^2 sum_ij(|data_ij - model_ij * g_i conj(g_j)| * wgts_ij)
+
+    and also a chisq_per_antenna which is the same sum but with fixed i. Also keeps track of the
+    number of unflagged observations that go into each chi^2 waterfal, both overall and per-antenna.
+
+    Arguments:
+        data: dictionary or DataContainer mapping baseline-pol keys like (1,2,'xx') to complex 2D
+            visibility data, with [0] axis time and [1] axis frequency.
+        model: dictionary or DataContainer mapping baseline-pol keys like (1,2,'xx') to complex 2D
+            visibility data, with [0] axis time and [1] axis frequency. Gains are multiplied by model
+            before comparing them to the data.
+        data_wgts: dictionary or DataContainer mapping baseline-pol keys like (1,2,'xx') to real
+            weights with [0] axis time and [1] axis frequency. Weights are interpeted as 1/sigma^2
+            where sigma is the noise on the data (but not necessarily the model if gains are provided).
+            Flags are be expressed as data_wgts equal to 0, so (~flags) produces binary weights.
+        gains: optional dictionary mapping ant-pol keys like (1,'x') to a waterfall of complex gains to
+            be multiplied into the model (or, equivalently, divided out from the data). Default: None,
+            which is interpreted as all gains are 1.0 (ie..e the data is already calibrated)
+        gain_flags: optional dictionary mapping ant-pol keys like (1,'x') to a boolean flags waterfall
+            with the same shape as the data. Default: None, which means no per-antenna flagging.
+        by_antpol: if True, chisq and nObs are dictionaries mapping antenna polarizations to numpy arrays.
```

Comment:
this kwarg name is a little confusing, but I don't have a better suggestion offhand. just something to think about. being a little more verbose with the kwarg name isn't a terrible thing...

Reply:
I'll make it `split_by_antpol`